Added CLI option internal_access for types (C#)

gvaish opened the pull request:
* `csharp_options`: Added `Options` to encapsulate generator options.
   Supported options for now - file_extension, base_namespace, access_level
* `{Blah}Generator`: Now accept `Options*` as parameter to constructor
* `csharp_generator.cc`: Parse and populate options
* `csharp_base_generator.cc`: `class_access_level` uses `file_options`,
     added `file_options()` getter for the options.

gvaish commented:
Addresses #1311

gvaish commented:
The `Options` type encapsulates command line options (`csharp_opt=`) and provides structured access.

Since these options are to be _used by generator_, they are passed in the constructor.

**Question**: Because it makes a sweeping change to all the constructors, looking for an alternative to minimize the diff.

cc: @jskeet @jtattermusch @anandolee (Authors from `csharp_generator.cc` log)

jskeet commented:
I'm not at all sure about this - I would at least want to see the Options change separately from adding a new option for access_level. We've tried really hard to keep options to the bare minimum - and the fact that there are in-proto ones as well as command-line ones adds to the confusion. That said, I can see the benefits in being able to make the access internal - that's what we do for the descriptor protos already, for example.

I'd be interested to hear what others think though.

Note that it's a UK four-day weekend, so I may well not be contributing much to the discussion over hat period.

gvaish commented:
@jskeet 

`this` = the `access_level` option or the `Options` class?

For the latter, I came up with this solution based on two observations:

1. `cpp` compiler has it too: https://github.com/google/protobuf/blob/master/src/google/protobuf/compiler/cpp/cpp_options.h
2. We already got 4 options (including `access_level`): https://github.com/gvaish/protobuf/blob/master/src/google/protobuf/compiler/csharp/csharp_options.h#L48-L55

One definitive alternative is to add `std::string` paramter to `Enum`, `Message` and `ReflectionClass` generators.

Regarding your concern on the number of options - Agree: Yes and No.

Yes. Agree that there are multiple sets of options - per-descriptor and per-code-generation. and is already confusing (to me too).

No. But then, one is core to entity description and other to the developer generating code.

I have a thought - may be have the options in a file, say `protoc.properties` (or whatever you want to call it). That way there can be a global set of properties that can be reused. The file can be searched in `./`, `~/`, `$PATH` folders in that order or overridden through a CLI option.

Thoughts?

jskeet commented:
Thanks for your further thoughts. I'm pondering on this - the fact that this has come at the same time as options around enum casing and prefix removal is a little overwhelming, but it does allow us to look at things in the round.

There may not be much activity on this for the next 72 hours or so, but please don't think I've forgotten about it.

gvaish commented:
Ha ha. No worries. I'm monitoring the other thread as well.

And I can see the weight around naming conventions.

jskeet commented on src/google/protobuf/compiler/csharp/csharp_enum.cc:
```diff
@@ -49,8 +49,8 @@ namespace protobuf {
 
-EnumGenerator::EnumGenerator(const EnumDescriptor* descriptor) :
-    SourceGeneratorBase(descriptor->file()),
+EnumGenerator::EnumGenerator(const EnumDescriptor* descriptor, const Options* file_options) :
```
file_options is slightly misleading here - that sounds like it's a DescriptorProtos.FileOptions. I'd call it generator_options or just options.

gvaish replied:
The convention was a copy over from [cpp_generator](https://github.com/google/protobuf/blob/master/src/google/protobuf/compiler/cpp/cpp_generator.cc#L86).

`options` is just as good.

jskeet commented on src/google/protobuf/compiler/csharp/csharp_generator.cc:
```diff
@@ -28,7 +28,13 @@
 // (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+// Authors: Jan Tattermusch <jtattermusch@google.com>
```
I don't think we need the authors listed here at all.

gvaish replied:
Hmmm.... it was mostly from looking at other files. Some [files](https://github.com/google/protobuf/blob/master/src/google/protobuf/compiler/cpp/cpp_generator.cc#L31-L33) that I saw had them so I punched as well.

jskeet commented on src/google/protobuf/compiler/csharp/csharp_generator.cc:
```diff
@@ -71,15 +79,31 @@ bool Generator::Generate(
+
+  struct Options file_options;
+
+  //std::string file_extension = ".cs";
```
Please remove the commented out lines. It sounds like really the Options struct should know how to parse a `vector<pair<string, string>>`, rather than the code being here.

As noted before, `file_options` should be `generator_options` or similar.

gvaish replied:
Noted.

jskeet commented on src/google/protobuf/compiler/csharp/csharp_names.h:
```diff
@@ -90,9 +92,7 @@ string GetReflectionClassName(const FileDescriptor* descriptor);
 //    will contain the error message.
```
I think this *might* break GRPC; @jtattermusch, are you using this function? (It might be simpler just to extract the relevant options when calling this.)

jskeet commented on src/google/protobuf/compiler/csharp/csharp_options.h:
```diff
@@ -0,0 +1,64 @@
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// Author: Gaurav Vaish <gaurav.vaish@gmail.com>
```
Again, remove the Author part. Although a few files have this, I don't believe it's the normal style.

gvaish replied:
Ack.

jskeet commented on src/google/protobuf/compiler/csharp/csharp_options.h:
```diff
@@ -0,0 +1,64 @@
+struct Options {
+  Options() : file_extension(".cs"), base_namespace(""), generate_directories(false), access_level("public") {
+  }
+  // 1. file_extension: Defaults to ".cs"
```
Why the numbering and repetation of the name in these comments? I'd add comments such as:

```
// Extension to apply to message name to create a filename. Defaults to ".cs"
```

You could leave TODO comments for me to fill in the details if you want.

gvaish replied:
Ack. Will fill in.

jskeet commented on src/google/protobuf/compiler/csharp/csharp_source_generator_base.cc:
```diff
@@ -57,7 +59,12 @@ void SourceGeneratorBase::WriteGeneratedCodeAttributes(io::Printer* printer) {
 }
 
 std::string SourceGeneratorBase::class_access_level() {
```
With this change, we can remove the IsDescriptorProto call here, and just make sure we call protoc appropriately in the generation code. (We still need the IsDescriptorProto function for other things though.)

gvaish replied:
@jskeet Do you mean that the `descriptor.proto` can also be generated as `public` and `internal` as opposed to only `internal` now?

jskeet commented:
Having spoken with the team, we're in favour of the general change, but I'd really like to see it split into two commits as mentioned before. (I don't think that should be too hard - just remove everything about the access_level bit to start with.)

Given that the only sensible access levels are "public" or "internal", we might want to just make this a Boolean option, too, e.g. "internal_access=true".

gvaish commented:
@jskeet This is my understanding of the proposed changes:

1. Commit 1: Introduce `options`.
2. Commit 2: Refactor to use the `options` - not just with access level (`internal_access`) but also with other options (`base_namespace` etc)

jskeet commented:
Yes, that's right. Ideally without changing csharp_names which is exported
and used by grpc, I think.
On 30 Mar 2016 7:10 p.m., "Gaurav Vaish" <notifications@github.com> wrote:

> @jskeet <https://github.com/jskeet> This is my understanding of the
> proposed changes:
>
>    1. Commit 1: Introduce options.
>    2. Commit 2: Refactor to use the options - not just with access level (
>    internal_access) but also with other options (base_namespace etc)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/google/protobuf/pull/1349#issuecomment-203560080>
>

gvaish commented:
Great. Will also revert `csharp_names` changes - will flatten the option values.

Didn't know about the `GRPC` integration. Because it didn't break `protoc` CLI, I considered it all green.

gvaish commented:
@jskeet First commit is in.

Let me know if you're ok to have this PR updated for the 2nd commit or fire another PR after this merge?

The 2nd commit will be a change in `csharp_generator_base.cc` in `SourceGeneratorBase::class_access_level()` to use `options`

jskeet commented:
Maybe it would be cleanest to have this PR *just* introduce the options type, then do the next step in another PR. To that end, could you remove the `internal_access` part of the options in this commit? It'll make it cleaner in that the next commit/PR will be "everything needed to add internal access". Sorry for the faff :)

gvaish commented:
That was some faff... I should have seen that coming ;-)

Anyway, here it is.

Once merged, I'll send out the next PR.

jskeet commented on src/gen.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
My guess is you didn't intend to include this?

gvaish replied:
Oh yeah. My bad for sure.

jskeet commented on src/google/protobuf/compiler/csharp/csharp_enum.cc:
```diff
@@ -49,8 +49,8 @@ namespace protobuf {
 
-EnumGenerator::EnumGenerator(const EnumDescriptor* descriptor) :
-    SourceGeneratorBase(descriptor->file()),
+EnumGenerator::EnumGenerator(const EnumDescriptor* descriptor, const Options* options) :
```
Looks like this could be a `const Options& options` (it's never going to be null) but I understand the desire to be consistent with the descriptor parameter. We might want to do a thorough cleanup at some point...

gvaish replied:
I thought of the `&` option but kept it consistent.

Let's keep the cleanup for later

jskeet commented on src/google/protobuf/compiler/csharp/csharp_options.h:
```diff
@@ -0,0 +1,60 @@
+// Protocol Buffers - Google's data interchange format
```
This needs to be added into `protobuf/cmake/extract_includes.bat.in` and `src/Makefile.am`, I believe - I think that's why the Travis build is failing.

gvaish replied:
Makes sense.

gvaish replied:
Or should they be in `nobase_include_HEADERS`?

jskeet replied:
The cpp options are under `libprotoc_la_SOURCES` - I think that's probably appropriate for the C# ones too.

jskeet commented:
Seems reasonable to me apart from the extra shell file and not adding the .h file to other build files. Happy to merge when those have been fixed.

grpc-jenkins commented:
Can one of the admins verify this patch?

gvaish commented:
Updates from the previous feedback:

1. Drop accidental check-in of `gen.sh`
2. Updated `src/Makefile.am` to include `csharp_options.h`
3. Updated `cmake/extract_includes.bat.in` to include `csharp_options.h`

Once done, I'll push out `internal_access` CLI parameter.

jskeet commented on src/Makefile.am:
```diff
@@ -468,7 +468,8 @@ libprotoc_la_SOURCES =                                         \
   google/protobuf/compiler/csharp/csharp_wrapper_field.cc      \
-  google/protobuf/compiler/csharp/csharp_wrapper_field.h
+  google/protobuf/compiler/csharp/csharp_wrapper_field.h       \
+  google/protobuf/compiler/csharp/csharp_options.h
```
Sorry to be even more of a pain, but all the files are listed alphabetically - could you move this line appropriately?

jskeet commented:
One final nit - then I'll just wait for Travis to go green, and merge. Depending on time, that may not be today (I'm in the UK) but fingers crossed...

haberman commented:
ok to test

gvaish commented:
@jskeet Couldn't do it yesterday. Crossed fingers failed.

But here you go.

jskeet commented:
Thanks! Will merge as soon as Travis goes green. I'll be sending out a PR tomorrow to add a bit more documentation and do a bit of line wrapping, but you've put up with enough requests from me :)

gvaish commented:
@jskeet Cool. I will create another PR for `internal_access` then.
